### PR TITLE
Added test script for uploading files in expenses and verifying it

### PR DESCRIPTION
**Pull request by KaranTikellu7**

### CODE REVIEW

File: aceinvoice-selenium-tests/src/Testcases/Expenses_Page/add_edit_archive_unarchive_expenses.java [hunk 1]
```diff
@@ -75,6 +77,22 @@ public void Add() throws InterruptedException
         OpenBrowser.driver.findElement(By.xpath("//select[@id='client_id']")).click();
         OpenBrowser.driver.findElement(By.xpath("//option[@value='240']")).click();
 
+        //Locating Choose file button.
+        WebElement ChooseFile_Button = OpenBrowser.driver.findElement(By.id("expense_attached_file_attributes_file"));
+
+        //Path of file to upload.
+        String File_to_upload= "C:\\Users\\HP\\Desktop\\aceinvoice-selenium-karan\\aceinvoice-selenium-tests\\DemoFiles_Folder\\DemoFile.txt";
```
Comment (prabhakarbattula) [inline aceinvoice-selenium-tests/src/Testcases/Expenses_Page/add_edit_archive_unarchive_expenses.java]: Create a folder with name `assets`.  Add the file to be uploaded into that assets folder.  Give a relative file reference to that path pointing to file in assets folder.

**Comment by prabhakarbattula** [comment]
@KaranTikellu7 

At the time of editing the expense change the file name and then verify the changed file name exists and the original file does not exist.

File: aceinvoice-selenium-tests/src/Testcases/Expenses_Page/add_edit_archive_unarchive_expenses.java [hunk 2]
```diff
@@ -75,6 +79,25 @@ public void Add() throws InterruptedException
         OpenBrowser.driver.findElement(By.xpath("//select[@id='client_id']")).click();
         OpenBrowser.driver.findElement(By.xpath("//option[@value='240']")).click();
 
+        //Locating Choose file button.
+        WebElement ChooseFile_Button = OpenBrowser.driver.findElement(By.id("expense_attached_file_attributes_file"));
+
+        //Path of file to upload.
+        String File_to_upload= "C:\\Users\\HP\\Desktop\\aceinvoice-selenium-karan\\aceinvoice-selenium-tests\\src\\Testcases\\Expenses_Page\\assets\\Demo-File.txt";
```
Comment (prabhakarbattula) [inline aceinvoice-selenium-tests/src/Testcases/Expenses_Page/add_edit_archive_unarchive_expenses.java]: Use relative path in line nos.86 and 116 as the one shown in screenshot.
image

**Comment by prabhakarbattula** [comment]
@KaranTikellu7 

Please update.

**Comment by prabhakarbattula** [comment]
@KaranTikellu7 

Please delete edit, archive and unarchive tests.
In add test case, condense file path variable into single line.

**Comment by KaranTikellu7** [comment]
@prabhakarbattula We have an issue with replacing the filename in relative path with variable. If we do so we assert the same file that we have uploaded and crosscheck with same file name store in variable , so the test will always pass.

**Comment by prabhakarbattula** [comment]
@KaranTikellu7 

The issue you mentioned is a different scenario where the two variables of an assertion are getting text from the same element which makes it a bad assertion and that will always pass.  That is different.

Here in this case, you can use as under
```
String file_name = "demo-expense.text";
String file_path = System.getProperty("user.dir") + "\\src\\assets\\" + file_name;  
```

File: aceinvoice-selenium-tests/src/Testcases/Expenses_Page/add_edit_archive_unarchive_expenses.java [hunk 3]
```diff
@@ -25,8 +28,13 @@
     String EditedVendor = "Demo-vendor-edited"+ RandomNumber;
     String VendorXpath = ("//span[contains(text(),'"+EditedVendor+"')]");
 
+    //Storing filenames into variable for uploading files.
+    String FileName= "Demo-File.txt";
+    String EditedFileName= "Edited-Demo-File.txt";
```
Comment (prabhakarbattula) [inline aceinvoice-selenium-tests/src/Testcases/Expenses_Page/add_edit_archive_unarchive_expenses.java]: Line 23 is no longer required.  Please remove.

File: aceinvoice-selenium-tests/src/Testcases/Expenses_Page/add_edit_archive_unarchive_expenses.java [hunk 4]
```diff
@@ -75,6 +83,22 @@ public void Add() throws InterruptedException
         OpenBrowser.driver.findElement(By.xpath("//select[@id='client_id']")).click();
         OpenBrowser.driver.findElement(By.xpath("//option[@value='240']")).click();
 
+        //Locating Choose file button.
+        WebElement ChooseFile_Button = OpenBrowser.driver.findElement(By.id("expense_attached_file_attributes_file"));
+
+        //Path of file to upload.
+        String FilePath = System.getProperty("user.dir")+"\\src\\Testcases\\Expenses_Page\\assets\\"+FileName;
```
Comment (prabhakarbattula) [inline aceinvoice-selenium-tests/src/Testcases/Expenses_Page/add_edit_archive_unarchive_expenses.java]: Please keep `assets` folder under `src` folder and not under `Expenses_Page` folder.

**Comment by prabhakarbattula** [comment]
@KaranTikellu7 

Please update.

**Pull request merged**